cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java

```diff
@@ -368,6 +368,31 @@ public int countActiveRuns(@Nullable Integer limit) {
     });
   }
 
+  @Override
+  public Map<ProgramId, Collection<RunRecordMeta>> getActiveRuns(Collection<ProgramId> programIds) {
+    return TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore appMetadataStore = getAppMetadataStore(context);
+      Map<ApplicationId, ApplicationMeta> metas =
+        appMetadataStore.getApplicationsForAppIds(
+          programIds.stream().map(ProgramId::getParent).collect(Collectors.toSet()));
+
+      // Get the active runs for programs that exist
+      Collection<ProgramId> existingProgramIds = new ArrayList<>();
+      for (ProgramId programId : programIds) {
+        // Ignore app or program that doesn't exist
+        ApplicationMeta appMeta = metas.get(programId.getParent());
+        if (appMeta == null) {
```

Comment:
can this condition be `appMeta == null || getExistingAppProgramSpecification(appMeta.getSpec(), programId) == null` ?